### DISCUSSION TITLE
DS18Alias doesn't work after reboot, if set on "USER_RULE" or "USER_BACKLOG"

### DISCUSSION THREAD

**Discussion by FerFemNemBem** [arendst/Tasmota#19875] [Q&A] [answered]
### PROBLEM DESCRIPTION
I have a self compiled tasmota V13.2.0 with the following defines enabled in my "my_user_config.h":
```
#define USE_RULES
#define USE_DS18x20
#define DS18x20_USE_ID_ALIAS

#define USER_RULE1 "ON System#Boot DO Backlog DS18Alias 7001203374828A28, Vorlauf; DS18Alias 510120335EE86028, Ruecklauf"
#define USER_BACKLOG "Rule1 1"
```

After startup the following MQTT-Message is sent:

`20:00:34.442 MQT: keller/heizung/warmwassersensor/SENSOR = {"Time":"2023-10-29T20:00:34","DS18B20-1":{"Id":"0120335EE860","Temperature":35.9},"DS18B20-2":{"Id":"01203374828A","Temperature":43.4},"TempUnit":"C"}`

Please recognize that the ID of the DS18B20 is another one (shorter) than the one from my rule (I will describe later why).
Inside the console I do:

`Backlog DS18Alias 0120335EE860, Ruecklauf; DS18Alias 01203374828A, Vorlauf`

These are the IDs from the MQTT-message (short IDs). This ends in:

```
20:02:11.063 CMD: Backlog DS18Alias 0120335EE860, Ruecklauf; DS18Alias 01203374828A, Vorlauf
20:02:11.132 MQT: keller/heizung/warmwassersensor/RESULT = {"DS18B20-1":{"Id":"510120335EE86028"},"DS18B20-2":{"Id":"7001203374828A28"}}
20:02:11.564 MQT: keller/heizung/warmwassersensor/RESULT = {"DS18B20-1":{"Id":"510120335EE86028"},"DS18B20-2":{"Id":"7001203374828A28"}}
```

Doing the same command but using the longer IDs, that are shown in the previous output:

`Backlog DS18Alias 510120335EE86028, Ruecklauf; DS18Alias 7001203374828A28, Vorlauf`

Results in:

```
20:03:14.601 CMD: Backlog DS18Alias 510120335EE86028, Ruecklauf; DS18Alias 7001203374828A28, Vorlauf
20:03:14.662 MQT: keller/heizung/warmwassersensor/RESULT = {"Ruecklauf":{"Id":"510120335EE86028"},"DS18B20-2":{"Id":"7001203374828A28"}}
20:03:15.101 MQT: keller/heizung/warmwassersensor/RESULT = {"Ruecklauf":{"Id":"510120335EE86028"},"Vorlauf":{"Id":"7001203374828A28"}}
```

This is correct and works - but will not survive a restart.

It doesn't matter if I use the short or the long IDs inside the rule
`"ON System#Boot DO Backlog..."`
or even if I use no rule and put the commands inside a
`#define USER_BACKLOG "DS18Alias 7001203374828A28, Vorlauf; DS18Alias 510120335EE86028, Ruecklauf"`

Also a "reset 2" after reflashing doesn't help.

### REQUESTED INFORMATION
_Make sure your have performed every step and checked the applicable boxes before submitting your issue. Thank you!_

- [ ] Read the [Contributing Guide and Policy](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md) and [the Code of Conduct](https://github.com/arendst/Tasmota/blob/development/CODE_OF_CONDUCT.md)
- [ ] Searched the problem in [issues](https://github.com/arendst/Tasmota/issues)
- [ ] Searched the problem in [discussions](https://github.com/arendst/Tasmota/discussions)
- [ ] Searched the problem in the [docs](https://tasmota.github.io/docs/FAQ)
- [ ] Searched the problem in the [chat](https://discord.gg/Ks2Kzd4)
- [ ] Device used (e.g., Sonoff Basic): _____
- [ ] Tasmota binary firmware version number used: _____
  - [ ] Pre-compiled
  - [ ] Self-compiled
- [ ] Flashing tools used: _____
- [ ] Provide the output of command: `Backlog Template; Module; GPIO 255`:
```lua
  Configuration output here:
20:18:51.194 MQT: keller/heizung/warmwassersensor/RESULT = {"NAME":"Warmwassersensor","GPIO":[1,1,1,1,1312,1,1,1,1,1,1,1,1,1],"FLAG":0,"BASE":18}
20:18:51.623 MQT: keller/heizung/warmwassersensor/RESULT = {"Module":{"0":"Warmwassersensor"}}
20:18:51.876 MQT: keller/heizung/warmwassersensor/RESULT = {"GPIO0":{"0":"None"},"GPIO1":{"0":"None"},"GPIO2":{"0":"None"},"GPIO3":{"0":"None"},"GPIO4":{"1312":"DS18x201"},"GPIO5":{"0":"None"},"GPIO9":{"0":"None"},"GPIO10":{"0":"None"},"GPIO12":{"0":"None"},"GPIO13":{"0":"None"},"GPIO14":{"0":"None"},"GPIO15":{"0":"None"},"GPIO16":{"0":"None"},"GPIO17":{"0":"None"}}
```
- [ ] If using rules, provide the output of this command: `Backlog Rule1; Rule2; Rule3`:
```lua
  Rules output here:
20:19:37.145 MQT: keller/heizung/warmwassersensor/RESULT = {"Rule1":{"State":"ON","Once":"OFF","StopOnError":"OFF","Length":100,"Free":411,"Rules":"ON System#Boot DO Backlog DS18Alias 7001203374828A28, Vorlauf; DS18Alias 510120335EE86028, Ruecklauf"}}

```
- [ ] Provide the output of this command: `Status 0`:
```lua
  STATUS 0 output here:
20:20:35.430 MQT: keller/heizung/warmwassersensor/STATUS = {"Status":{"Module":0,"DeviceName":"Warmwassersensor","FriendlyName":["Warmwassersensor"],"Topic":"keller/heizung/warmwassersensor","ButtonTopic":"0","Power":0,"PowerOnState":3,"LedState":1,"LedMask":"FFFF","SaveData":1,"SaveState":1,"SwitchTopic":"0","SwitchMode":[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],"ButtonRetain":0,"SwitchRetain":0,"SensorRetain":0,"PowerRetain":0,"InfoRetain":0,"StateRetain":0,"StatusRetain":0}}
20:20:35.452 MQT: keller/heizung/warmwassersensor/STATUS1 = {"StatusPRM":{"Baudrate":115200,"SerialConfig":"8N1","GroupTopic":"tasmotas","OtaUrl":"","RestartReason":"Software/System restart","Uptime":"0T00:20:13","StartupUTC":"2023-10-29T19:00:22","Sleep":250,"CfgHolder":5266,"BootCount":2,"BCResetTime":"2023-10-29T19:45:21","SaveCount":6,"SaveAddress":"F6000"}}
20:20:35.470 MQT: keller/heizung/warmwassersensor/STATUS2 = {"StatusFWR":{"Version":"13.2.0(tasmota)","BuildDateTime":"2023-10-29T19:37:45","Boot":31,"Core":"2_7_4_9","SDK":"2.2.2-dev(38a443e)","CpuFrequency":80,"Hardware":"ESP8266EX","CR":"536/699"}}
20:20:35.490 MQT: keller/heizung/warmwassersensor/STATUS3 = {"StatusLOG":{"SerialLog":2,"WebLog":2,"MqttLog":0,"SysLog":2,"LogHost":"192.168.178.72","LogPort":514,"SSId":["FFNB-WLAN",""],"TelePeriod":60,"Resolution":"558180C0","SetOption":["00208009","2805C80001000600001E5A00192800000000","000000C0","00006000","00046000","00000000"]}}
20:20:35.516 MQT: keller/heizung/warmwassersensor/STATUS4 = {"StatusMEM":{"ProgramSize":545,"Free":456,"Heap":26,"ProgramFlashSize":1024,"FlashSize":4096,"FlashChipId":"164020","FlashFrequency":40,"FlashMode":"DOUT","Features":["00000809","08400582","04000001","00000043","00000000","00000000","00008080","00000000","04000000","00000000"],"Drivers":"1,2,4,10,12,38,40,62,68","Sensors":"1,2,5"}}
20:20:35.538 MQT: keller/heizung/warmwassersensor/STATUS5 = {"StatusNET":{"Hostname":"keller-heizung-warmwassersensor","IPAddress":"192.168.178.61","Gateway":"192.168.178.1","Subnetmask":"255.255.255.0","DNSServer1":"192.168.178.1","DNSServer2":"0.0.0.0","Mac":"C8:2B:96:08:5D:9A","Webserver":2,"HTTP_API":1,"WifiConfig":4,"WifiPower":17.0}}
20:20:35.559 MQT: keller/heizung/warmwassersensor/STATUS6 = {"StatusMQT":{"MqttHost":"192.168.178.72","MqttPort":1883,"MqttClientMask":"DVES_%06X","MqttClient":"DVES_085D9A","MqttUser":"DVES_USER","MqttCount":1,"MAX_PACKET_SIZE":1200,"KEEPALIVE":30,"SOCKET_TIMEOUT":4}}
20:20:35.577 MQT: keller/heizung/warmwassersensor/STATUS7 = {"StatusTIM":{"UTC":"2023-10-29T19:20:35","Local":"2023-10-29T20:20:35","StartDST":"2023-03-26T02:00:00","EndDST":"2023-10-29T03:00:00","Timezone":99}}
20:20:35.597 MQT: keller/heizung/warmwassersensor/STATUS10 = {"StatusSNS":{"Time":"2023-10-29T20:20:35","Ruecklauf":{"Id":"0120335EE860","Temperature":34.8},"Vorlauf":{"Id":"01203374828A","Temperature":40.9},"TempUnit":"C"}}
20:20:35.615 MQT: keller/heizung/warmwassersensor/STATUS11 = {"StatusSTS":{"Time":"2023-10-29T20:20:35","Uptime":"0T00:20:13","UptimeSec":1213,"Heap":26,"SleepMode":"Dynamic","Sleep":250,"LoadAvg":10,"MqttCount":1,"Wifi":{"AP":1,"SSId":"FFNB-WLAN","BSSId":"3C:A6:2F:B4:59:D0","Channel":6,"Mode":"11n","RSSI":44,"Signal":-78,"LinkCount":1,"Downtime":"0T00:00:06"}}}

```
- [ ] Set `weblog` to 4 and then, when you experience your issue, provide the output of the Console log:
```lua
  Console output here:


```

### TO REPRODUCE
_Steps to reproduce the behavior:_

### EXPECTED BEHAVIOUR
_A clear and concise description of what you expected to happen._

### SCREENSHOTS
_If applicable, add screenshots to help explain your problem._

### ADDITIONAL CONTEXT
_Add any other context about the problem here._

**(Please, remember to close the issue when the problem has been addressed)**

**Comment by FerFemNemBem** [comment]
As I can reproduce this on several devices - is there anything I'm doing wrong?
Why was this moved to "discussion" from "issues". Is this feature working in general?

**Reply by sfromis** [reply]
Moved because it is not about a bug, but about the feature being a bit hard to understand.

Yes, you need to use the full 64-bit `Id` value in the `DS18Alias` command, instead of the shorter 48-bit one listed when publishing the sensor payloads. You can get the 64-bit values by entering the command without arguments, or just each time you use the command.

As the `DS18Alias` assignments are not persistent, the command needs to be repeated each boot. This is as designed. Therefore it cannot work in `USER_BACKLOG` which is only executed once, in the first initialization boot. Not on restarts thereafter.

It works fine when you have the `DS18Alias` command(s) in a `System#boot` rule, or in `autoexec.bat` (if you have a file system). To embed in a build, you can use `USER_RULE1` for the commands, and enable the rule set in `USER_BACKLOG`. This works as rule sets are persistent.

**Reply by sfromis** [reply]
Of course, it would've helped to read the docs about the feature. https://tasmota.github.io/docs/DS18x20/#sensor-aliases

**Reply by FerFemNemBem** [reply]
As mentioned - it also doesn't work if I use the full 64-bit ID:
`#define USER_RULE1 "ON System#Boot DO Backlog DS18Alias 7001203374828A28, Vorlauf; DS18Alias 510120335EE86028, Ruecklauf`
Please point me to what I'm doing wrong.
Thanks in advance,
FFNB

**Reply by sfromis** [reply, marked as answer]
Your rule is invalid (missing `endon`), and you did not mention if you enabled the rule set too. Also, that way can only work if you do a full reset to initial defaults after upgrading. I expect that the lacking `"` at the end is a copy/paste issue.

**Reply by FerFemNemBem** [reply]
I'm really, really sorry! Because the copy/paste from my `USER_BACKLOG` to the `USER_RULE1` I've forgot the `ENDON`. 
All is working as expected using the adapted RULE and doing a `reset 2` afterwards!

Thanks a lot!
FFNB
